services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/service/MongoDbBasedTenantService.java

```diff
@@ -0,0 +1,630 @@
+     *                  that has been configured for the tenant.
+     * @param span      The active OpenTracing span for this operation. It is not to be closed in this method!
+     *                  An implementation should log (error) events on this span and it may set tags and use this span as the
+     *                  parent for any spans created in this method.
+     * @return Future TenantResult with {@code HttpURLConnection.HTTP_OK} if tenant found, {@code HttpURLConnection.HTTP_NOT_FOUND} else.
+     * @throws NullPointerException if {@code subjectDn} is {@code null}.
+     */
+    @Override
+    public Future<TenantResult<JsonObject>> get(final X500Principal subjectDn, final Span span) {
+        Objects.requireNonNull(subjectDn);
+
+        return getForCertificateAuthority(subjectDn, span);
+
+    }
+
+    /**
+     * Finds an existing tenant by the subject DN of its configured certificate authority.
+     *
+     * @param subjectDn the subject DN to find the tenant with.
+     * @param span      The active OpenTracing span for this operation. It is not to be closed in this method!
+     *                  An implementation should log (error) events on this span and it may set tags and use this span as the
+     *                  parent for any spans created in this method.
+     * @return Future TenantResult with {@code HttpURLConnection.HTTP_OK} if tenant found, {@code HttpURLConnection.HTTP_NOT_FOUND} else.
+     */
+    private Future<TenantResult<JsonObject>> getForCertificateAuthority(final X500Principal subjectDn, final Span span) {
+        if (subjectDn == null) {
+            TracingHelper.logError(span, "missing subject DN");
+            return Future.succeededFuture(TenantResult.from(HttpURLConnection.HTTP_BAD_REQUEST));
+        } else {
+            final Future<Map.Entry<String, Versioned<Tenant>>> tenantsFound = getByCa(subjectDn);
+            return tenantsFound.compose(successTenantRead -> {
+                if (successTenantRead == null) {
+                    TracingHelper.logError(span, "no tenant found for subject DN");
+                    return Future.succeededFuture(TenantResult.from(HttpURLConnection.HTTP_NOT_FOUND));
+                } else {
+                    return Future.succeededFuture(TenantResult.from(
+                            HttpURLConnection.HTTP_OK,
+                            DeviceRegistryUtils.convertTenant(successTenantRead.getKey(), successTenantRead.getValue().getValue(), true),
+                            DeviceRegistryUtils.getCacheDirective(config.getCacheMaxAge())));
+                }
+            });
+        }
+    }
+
+    /**
+     * Fetches tenant by subject DN.
+     *
+     * @param subjectDn the subject DN.
+     * @return A Future Map consisting of pair of a tenant id and a versioned tenant.
+     */
+    private Future<Map.Entry<String, Versioned<Tenant>>> getByCa(final X500Principal subjectDn) {
```

Comment:
This method is also available in file based device registry. Duplication could be avoided by pulling up this method to `DeviceRegistryUtils` in base module.

Reply:
The problem with this is that the method calls another local method and the extraction of the method becomes more complex.
Proposal: Extract this method along others in a refactor PR later.